transactor/src/main/scala/io/mediachain/transactor/Copycat.scala

```diff
@@ -0,0 +1,186 @@
+package io.mediachain.transactor
+
+import scala.collection.mutable.{Set => MSet, HashSet => MHashSet, 
+                                 Map => MMap, HashMap => MHashMap}
+
+import io.atomix.copycat.{Command, Query}
+import io.atomix.copycat.server.{Commit, StateMachine, Snapshottable}
+import io.atomix.copycat.server.session.{ServerSession, SessionListener}
+import io.atomix.copycat.server.storage.snapshot.{SnapshotReader, SnapshotWriter}
+
+import cats.data.Xor
+
+object Copycat {
+  import io.mediachain.transactor.Types._
+  
+  case class JournalInsert(
+    record: CanonicalRecord
+  ) extends Command[Xor[JournalError, CanonicalEntry]]
+
+  case class JournalUpdate( 
+    ref: Reference,
+    cell: ChainCell
+  ) extends Command[Xor[JournalError, ChainEntry]]
+  
+  case class JournalLookup(
+    ref: Reference
+  ) extends Query[Option[Reference]]
+  
+  case class JournalCommitEvent(entry: JournalEntry)
+
+  class JournalStateMachine(
+    val datastore: Datastore
+  ) extends StateMachine with Snapshottable with SessionListener {
+    sealed abstract class CanonicalReference {
+      def chain: Option[Reference]
+    }
+    case class EntityReference(chain: Option[Reference]) extends CanonicalReference
+    case class ArtefactReference(chain: Option[Reference]) extends CanonicalReference
+    
+    private var seqno: BigInt = 0
+    private var index: MMap[Reference, CanonicalReference] = new MHashMap    
+    private var clients: MSet[ServerSession] = new MHashSet // this wanted to be called sessions
+
+    
+    private def CommitError(what: String) = Xor.left(JournalCommitError(what))
+    
+    // Journal Interface
+    def insert(commit: Commit[JournalInsert]): Xor[JournalError, CanonicalEntry] = {
+      try {
+        insertRecord(commit.operation)
+      } finally {
+        commit.release()
+      }
+    }
+    
+    private def insertRecord(cmd: JournalInsert): Xor[JournalError, CanonicalEntry] = {
+      val rec = cmd.record
+      val ref = datastore.put(rec)
+      index.get(ref) match {
+        case Some(_) => CommitError("duplicate insert")
+        case None => {
+          rec match {
+            case Entity(_) => {
+              index += (ref -> EntityReference(None))
+            }
+            case Artefact(_) => {
+              index += (ref -> ArtefactReference(None))
+            }
+          }
+          
+          val entry = CanonicalEntry(nextSeqno(), ref)
+          publishCommit(entry)
+          Xor.right(entry)
+        }
+      }
+    }
+
+    def update(commit: Commit[JournalUpdate]): Xor[JournalError, ChainEntry] = {
+      try {
+        updateChain(commit.operation)
+      } finally {
+        commit.release()
+      }
+    }
+    
+    private def updateChain(cmd: JournalUpdate): Xor[JournalError, ChainEntry] = {
+      val ref = cmd.ref
+      val cell = cmd.cell
+      index.get(ref) match {
```

Comment:
Likewise, I would try to write this in terms of tuples instead of nesting, since the branches are quite unbalanced.

You can also embed your `if` into the case statement as a guard, see https://www.safaribooksonline.com/library/view/scala-cookbook/9781449340292/ch03s14.html

Scala is all about pattern matching!